core/domain/exp_services.py

```diff
@@ -586,6 +588,270 @@ def apply_change_list(exploration_id, change_list):
         raise e
 
 
+def update_states_version_history(
+    states_version_history, change_list, old_states, new_states,
+    current_version, committer_id):
+    """Updates the version history of each state at a particular version
+    of an exploration.
+
+    Args:
+        states_version_history: dict(str, StateVersionHistory). The version
+            history data of each state in the previous version of the
+            exploration.
+        change_list: list(ExplorationChange). A list of changes introduced in
+            this commit.
+        old_states: dict(str, State). The states in the previous version of
+            the exploration.
+        new_states: dict(str, State). The states in the current version of
+            the exploration.
+        current_version: int. The latest version of the exploration.
+        committer_id: str. The id of the user who made the commit.
+
+    Returns:
+        states_version_history: dict(str, StateVersionHistory). The updated
+        version history data of each state.
+    """
+    exp_versions_diff = exp_domain.ExplorationVersionsDiff(change_list)
+    prev_version = current_version - 1
+    old_states_dict = copy.deepcopy({
+        state_name: state.to_dict()
+        for (state_name, state) in old_states.items()
+    })
+    new_states_dict = copy.deepcopy({
+        state_name: state.to_dict()
+        for (state_name, state) in new_states.items()
+    })
+
+    # Firstly, delete the states from the state version history which were
+    # deleted during this commit.
+    for state_name in exp_versions_diff.deleted_state_names:
+        del states_version_history[state_name]
+
+    # Now, add the states which were newly added during this commit. The
+    # version history of these states are initialized as None because they
+    # were newly added and have no 'previously edited version'.
+    for state_name in exp_versions_diff.added_state_names:
+        states_version_history[state_name] = (
+            state_domain.StateVersionHistory(None, None, committer_id))
+
+    # Now, handle the updation of version history of states which were renamed.
+    # Firstly, we need to clean up the exp_versions_diff.old_to_new_state_names
+    # dict from the state names which are not effectively changed. For example,
+    # if a state was renamed from state_1 to state_2 and then from state_2 to
+    # state_1 in the same commit, then there is no effective change in state
+    # name and we need to clear them from this dict.
```

Comment:
Can there be a case where 2 (or more) state names are interchanged? If so say you rename state_1 to state_2 and rename an already existing state_2 to state_1? In which case the below logic might run into problems right?

Reply:
Thanks for suggesting this case, I have taken this into account now and did some refactor. Now, this case is handled properly. I have also added a backend test to ensure the same.